### PR TITLE
feat: add visionOS support

### PR DESCRIPTION
**Pull request by okwasniewski**
Hey,

This PR introduces visionOS windowed rendering for Babylon Native.

### Recording

https://github.com/user-attachments/assets/8cf0104b-3267-4fa7-9361-7351637cf114

## TODO
- [x] Merge bgfx visionOS support: https://github.com/bkaradzic/bgfx/pull/3289
  - [x] https://github.com/bkaradzic/bgfx/pull/3326
  - [x] https://github.com/bkaradzic/bgfx/pull/3325
- [x] Sync changes from `bkaradzic/bgfx` to `BabylonJS/bgfx`: https://github.com/BabylonJS/bgfx/pull/38
- [x] Sync [bgfx.cmake](https://github.com/BabylonJS/bgfx.cmake) repo: https://github.com/BabylonJS/bgfx.cmake/pull/77

NativeXr plugin integration will be a separate PR.

### CODE REVIEW

**Comment by okwasniewski** [comment]
I'm not sure if the latest BGFX changes make Win32 fail or if it's something else... Probably it would be best to separate bgfx bump

**Comment by bghgary** [comment]
> I'm not sure if the latest BGFX changes make Win32 fail or if it's something else... Probably it would be best to separate bgfx bump

It's probably bgfx. We normally verify that bgfx is working before submitting a PR to the bgfx fork. Here is one of the errors:

```
##[error]buildUWP_arm64\_deps\bgfx.cmake-src\bgfx\src\shader_spirv.cpp(43,73): Error C2065: 'LiteralNumber': undeclared identifier
    98>D:\a\1\s\buildUWP_arm64\_deps\bgfx.cmake-src\bgfx\src\shader_spirv.cpp(43,73): error C2065: 'LiteralNumber': undeclared identifier [D:\a\1\s\buildUWP_arm64\_deps\bgfx.cmake-build\bgfx.vcxproj]
         (compiling source file 'CMakeFiles/bgfx.dir/Unity/unity_2_cxx.cxx')
```

**Comment by okwasniewski** [comment]
Okay looks like this change fixed build issues: 


```diff
diff --git a/src/shader_spirv.cpp b/src/shader_spirv.cpp
index dc4f12282..185fdfd02 100644
--- a/src/shader_spirv.cpp
+++ b/src/shader_spirv.cpp
@@ -17,12 +17,8 @@ namespace bgfx
 #define SPV_OPERAND_7(_a0, _a1, _a2, _a3, _a4, _a5, _a6) SPV_OPERAND_1(_a0), SPV_OPERAND_6(_a1, _a2, _a3, _a4, _a5, _a6)
 #define SPV_OPERAND_8(_a0, _a1, _a2, _a3, _a4, _a5, _a6, _a7) SPV_OPERAND_1(_a0), SPV_OPERAND_7(_a1, _a2, _a3, _a4, _a5, _a6, _a7)
 #define SPV_OPERAND_9(_a0, _a1, _a2, _a3, _a4, _a5, _a6, _a7, _a8) SPV_OPERAND_1(_a0), SPV_OPERAND_8(_a1, _a2, _a3, _a4, _a5, _a6, _a7, _a8)
-#if BX_COMPILER_MSVC
-// Workaround MSVS bug...
-#	define SPV_OPERAND(...) { BX_MACRO_DISPATCHER(SPV_OPERAND_, __VA_ARGS__) BX_VA_ARGS_PASS(__VA_ARGS__) }
-#else
-#	define SPV_OPERAND(...) { BX_MACRO_DISPATCHER(SPV_OPERAND_, __VA_ARGS__)(__VA_ARGS__) }
-#endif // BX_COMPILER_MSVC
+#define SPV_OPERAND(...) { BX_MACRO_DISPATCHER(SPV_OPERAND_, __VA_ARGS__)(__VA_ARGS__) }
+
 #define _ Count
 
 	bool isDebug(SpvOpcode::Enum _opcode)
```

Now it's the Validation test: 

```
##[error]Cmd.exe exited with code '-1'.
```

**Comment by okwasniewski** [comment]
Hey, FYI we will first make windowed rendering work in this PR and then create a separate one with the NativeXr plugin.

**Comment by okwasniewski** [comment]
This PR should also fix simulator builds for both iOS and visionOS 

CleanShot 2024-08-14 at 16 16 00@2x

**Comment by okwasniewski** [comment]
Code-wise PR is ready to review, I still need to get the latest sync merged: https://github.com/BabylonJS/bgfx/pull/40

File: Apps/Playground/CMakeLists.txt
Comment (ryantrem) [inline Apps/Playground/CMakeLists.txt:72]: What are the warnings/errors that make things otherwise not work for simualtors?

File: Apps/Playground/CMakeLists.txt
Comment (ryantrem) [inline Apps/Playground/CMakeLists.txt:172]: Is this newly needed on iOS?

File: Apps/Playground/CMakeLists.txt
Comment (ryantrem) [inline Apps/Playground/CMakeLists.txt:207]: These are indented too far, really through me off for a sec. I assume this is just a copy/paste of the same pre-existing formatting issue from the iOS section.
```suggestion
        set(CMAKE_Swift_COMPILER_FORCED TRUE)
        set(CMAKE_Swift_LANGUAGE_VERSION 5.0)
        enable_language(Swift)
```

File: CMakeLists.txt
Comment (ryantrem) [inline CMakeLists.txt:37]: Is this change needed because we rely on this toolchain for visionOS as well? Is there a release tag that includes visionOS support, or we have to just grab a specific commit?

File: CMakeLists.txt
Comment (ryantrem) [inline CMakeLists.txt:26]: I know this is already on your radar, but just putting a comment so we don't forget to update this once the bgfx changes are merged.

File: CMakeLists.txt
Comment (ryantrem) [inline CMakeLists.txt:157]: Were there actual changes that needed this?

**Comment by ryantrem** [comment]
Is the build error related to the bgfx update?

File: CMakeLists.txt
Comment (okwasniewski) [reply CMakeLists.txt:37]: They recently released a new version which Im going to bump in another PR just to make sure other builds are still working as they should.

File: CMakeLists.txt
Comment (okwasniewski) [reply CMakeLists.txt:157]: This is also going to be added in another PR (bgfx bump) to separate changes

Comment (okwasniewski) [reply CMakeLists.txt:157]: https://github.com/BabylonJS/BabylonNative/pull/1408

File: CMakeLists.txt
Comment (okwasniewski) [reply CMakeLists.txt:26]: I will remove it once https://github.com/BabylonJS/bgfx/pull/40 gets merged. Let's leave this conversation open

File: Apps/Playground/CMakeLists.txt
Comment (okwasniewski) [reply Apps/Playground/CMakeLists.txt:72]: It was treating internal warnings as errors, like this:

CleanShot 2024-08-22 at 17 11 52@2x

File: Apps/Playground/CMakeLists.txt
Comment (okwasniewski) [reply Apps/Playground/CMakeLists.txt:172]: I don't think it's new. This is an additional fix for the simulator builds without it the simulator doesn't link built-in libraries. 

CleanShot 2024-08-27 at 13 53 13@2x

File: Apps/Playground/CMakeLists.txt
Comment (okwasniewski) [reply Apps/Playground/CMakeLists.txt]: Fixed for both sections

**Comment by okwasniewski** [comment]
Hey @ryantrem @CedricGuillemet this one is ready for re-review!

File: Apps/Playground/CMakeLists.txt
Comment (CedricGuillemet) [inline Apps/Playground/CMakeLists.txt:70]: Do you have some infos on errors/ warnings here ?

Comment (okwasniewski) [reply Apps/Playground/CMakeLists.txt:70]: Yes, https://github.com/BabylonJS/BabylonNative/pull/1384#discussion_r1727289518

**Review by docEdub — COMMENTED** [review]
Thanks for this PR. I'm getting ready to test on-device, now. Please add a link in README.md to the new VisionOS section in BUILDING.md.

**Comment by docEdub** [comment]
I'm getting a bgfx build error
```
[...]/build/visionOS/_deps/bgfx.cmake-src/bgfx/src/renderer_mtl.mm:3528:5: error: 'MTKView' is unavailable: not available on visionOS
                                MTKView *view = (MTKView *)_nwh;
                                ^
```

**Comment by okwasniewski** [comment]
@docEdub 

Which version of CMake are you using? Can you try with at least 3.29.0? 

And also just to make sure did you use this command to generate project?

```
cmake -B build/visionOS -G Xcode -D VISIONOS=ON
```

**Comment by ryantrem** [comment]
@okwasniewski in the PR description you have a link to a closed bgfx.cmake PR. Will you update that to point to the merged PR(s) so we have a good record of all the related changes across repos?

**Comment by okwasniewski** [comment]
> @okwasniewski in the PR description you have a link to a closed bgfx.cmake PR. Will you update that to point to the merged PR(s) so we have a good record of all the related changes across repos?

Fixed

**Comment by docEdub** [comment]
> @docEdub
> 
> Which version of CMake are you using? Can you try with at least 3.29.0?
> 
> And also just to make sure did you use this command to generate project?
> 
> ```
> cmake -B build/visionOS -G Xcode -D VISIONOS=ON
> ```

I'm have CMake 3.29.3 installed and I'm using `cmake -B build/visionOS -G Xcode -D VISIONOS=ON` to generate the Xcode project.

**Comment by docEdub** [comment]
> I'm getting a bgfx build error
> 
> ```
> [...]/build/visionOS/_deps/bgfx.cmake-src/bgfx/src/renderer_mtl.mm:3528:5: error: 'MTKView' is unavailable: not available on visionOS
>                                 MTKView *view = (MTKView *)_nwh;
>                                 ^
> ```

Is `MTKView` expected to be available? ...or is this section of renderer_mtl.mm supposed to be skipped for VisionOS builds?

**Comment by okwasniewski** [comment]
I just did another test of installing fresh deps: 

```
rm -rf build/visionOS
```
And then: 

```
cmake -B build/visionOS -G Xcode -D VISIONOS=ON
```

And everything worked fine: 
CleanShot 2024-08-28 at 16 41 38@2x

**Comment by docEdub** [comment]
> I just did another test of installing fresh deps:
> 
> ```
> rm -rf build/visionOS
> ```
> 
> And then:
> 
> ```
> cmake -B build/visionOS -G Xcode -D VISIONOS=ON
> ```
> 
> And everything worked fine: CleanShot 2024-08-28 at 16 41 38@2x

Which visionOS SDK are you using? I'm using visionOS SDK 1.2.

**Comment by okwasniewski** [comment]
Oh okay I got the issue, I was running on visionOS 2.0 and switched to older Xcode for simulator build and got the same issue as you did.

 Damn I guess I need to submit one more PR to bgfx. Going to work on this tomorrow

**Comment by docEdub** [comment]
> > MTKView *view = (MTKView *)_nwh;
> 
> It's supposed to be skipped, in that file (`renderer_mtl.mm`) from BGFX do you have any signs of `BX_PLATFORM_VISIONOS` macro? Maybe the dependency got cached somehow?

Ok, ya, the line causing the build error is outside of the `BX_PLATFORM_VISIONOS` macros:

```
	void SwapChainMtl::init(void* _nwh)
	{
#if BX_PLATFORM_VISIONOS
		NSObject* nvh = (NSObject*)_nwh;
		m_useLayerRenderer = ![nvh isKindOfClass:[CAMetalLayer class]];
		if (m_useLayerRenderer)
		{
			cp_layer_renderer_t layerRenderer = (cp_layer_renderer_t)_nwh;
			m_layerRenderer = layerRenderer;
			m_layerRendererConfiguration = cp_layer_renderer_get_configuration(m_layerRenderer);
			
			if (cp_layer_renderer_configuration_get_layout(m_layerRendererConfiguration) == cp_layer_renderer_layout_dedicated) {
				BX_WARN(false, "Dedicated layer renderer layout is not supported.");
			}
			
			retain(m_layerRendererConfiguration);
			retain(m_layerRenderer);
		}
		else
#endif // BX_PLATFORM_VISIONOS
		{
			if (m_metalLayer)
			{
				release(m_metalLayer);
			}
			
			if (NULL != NSClassFromString(@"MTKView") )
			{
build error -->		MTKView *view = (MTKView *)_nwh;
				
				if (NULL != view
					&&  [view isKindOfClass:NSClassFromString(@"MTKView")])
				{
					m_metalLayer = (CAMetalLayer *)view.layer;
				}
			}
			
			if (NULL != NSClassFromString(@"CAMetalLayer") )
			{
				if (NULL == m_metalLayer)
#	if BX_PLATFORM_IOS || BX_PLATFORM_VISIONOS
				{
					CAMetalLayer* metalLayer = (CAMetalLayer*)_nwh;
```

Do you have the same implementation for `SwapChainMtl`?

**Comment by okwasniewski** [comment]
@docEdub Can you try if adding ifdefs to this piece of code make bgfx compile? Im hoping you could test out Babylon Native while we wait for this PR to get merged into bgfx 

```objc
#if !BX_PLATFORM_VISIONOS
			if (NULL != NSClassFromString(@"MTKView") )
			{
				MTKView *view = (MTKView *)_nwh;
				
				if (NULL != view
					&&  [view isKindOfClass:NSClassFromString(@"MTKView")])
				{
					m_metalLayer = (CAMetalLayer *)view.layer;
				}
			}
#endif
```

**Comment by okwasniewski** [comment]
Here is the PR which should fix it: https://github.com/bkaradzic/bgfx/pull/3347 going to do sync one more time once it merges

**Comment by docEdub** [comment]
> @docEdub Can you try if adding ifdefs to this piece of code make bgfx compile? Im hoping you could test out Babylon Native while we wait for this PR to get merged into bgfx
> 
> ```objc
> #if !BX_PLATFORM_VISIONOS
> 			if (NULL != NSClassFromString(@"MTKView") )
> 			{
> 				MTKView *view = (MTKView *)_nwh;
> 				
> 				if (NULL != view
> 					&&  [view isKindOfClass:NSClassFromString(@"MTKView")])
> 				{
> 					m_metalLayer = (CAMetalLayer *)view.layer;
> 				}
> 			}
> #endif
> ```

Yep, that gets it building and it's working on device. Nice!

**Comment by ryantrem** [comment]
Looks like we're good to go on this one?

**Review by ryantrem — APPROVED** [review]

**Review by docEdub — APPROVED** [review]
👍 LGTM. Tested on device with most recent changes and it works as expected.

**Comment by bghgary** [comment]
Congrats @okwasniewski (and @matthargett) on getting this in! Thanks for the contribution!

**Comment by matthargett** [comment]
Thanks for your partnership (and patience)! Next up is getting Babylon React Native to support React Native visionOS, then immersive mode and 6DOF head tracking 🥽

**Pull request merged**
